### PR TITLE
Do not propagate profile changes of shadow-banned users into rooms

### PR DESCRIPTION
**Pull request by clokep**
This is another part of #8028 that is split out. It avoids messages like "foo changed their display name to bar" from appearing for shadow-banned users.

### CODE REVIEW

**Comment by clokep** [comment]
This does not seem to help with changing the nick for a particular room though, which seems to come through a different endpoint.

**Comment by anoadragon453** [comment]
> This does not seem to help with changing the nick for a particular room though, which seems to come through a different endpoint.

That's the result of a user sending another `m.room.member` state event of type `join` into the room, via [PUT http://matrix.example.com/_matrix/client/r0/rooms/!room:example.com/state/m.room.member/@user:example.com](https://matrix.org/docs/spec/client_server/latest#put-matrix-client-r0-rooms-roomid-state-eventtype-statekey).

File: synapse/handlers/profile.py [hunk 1]
```diff
@@ -213,8 +214,14 @@ async def get_avatar_url(self, target_user):
...
         self, target_user, requester, new_avatar_url, by_admin=False
     ):
-        """target_user is the user whose avatar_url is to be changed;
-        auth_user is the user attempting to make this change."""
+        """Set a new avatar URL for a user.
+
+        Args:
+            target_user (UserID): the user whose displayname is to be changed.
+            requester (Requester): The user attempting to make this change.
+            new_displayname (str): The displayname to give this user.
```
Comment (anoadragon453) [inline synapse/handlers/profile.py:222]:
```suggestion
            new_avatar_url (str): The avatar URL to give this user.
```

File: synapse/handlers/profile.py [hunk 2]
```diff
@@ -213,8 +214,14 @@ async def get_avatar_url(self, target_user):
     async def set_avatar_url(
         self, target_user, requester, new_avatar_url, by_admin=False
     ):
-        """target_user is the user whose avatar_url is to be changed;
-        auth_user is the user attempting to make this change."""
+        """Set a new avatar URL for a user.
+
+        Args:
+            target_user (UserID): the user whose displayname is to be changed.
```
Comment (anoadragon453) [inline synapse/handlers/profile.py:220]:
```suggestion
            target_user (UserID): the user whose avatar URL is to be changed.
```

File: tests/rest/client/test_shadow_banned.py [hunk 3]
```diff
@@ -16,23 +16,13 @@
...
-class RoomTestCase(unittest.HomeserverTestCase):
-    servlets = [
-        synapse.rest.admin.register_servlets_for_client_rest_resource,
-        directory.register_servlets,
-        login.register_servlets,
-        room.register_servlets,
-        room_upgrade_rest_servlet.register_servlets,
-    ]
-
+class _ShadowBannedBase(unittest.HomeserverTestCase):
```
Comment (anoadragon453) [inline tests/rest/client/test_shadow_banned.py:25]: Elegant way to break the classes up :+1:

File: tests/rest/client/test_shadow_banned.py [hunk 4]
```diff
@@ -176,3 +178,34 @@ def test_upgrade(self):
...
+        self.assertEquals(200, channel.code, channel.result)
+        self.assertEqual(channel.json_body, {})
+
+        # The user's display name should be updated.
+        request, channel = self.make_request(
+            "GET", "/profile/%s/displayname" % (self.banned_user_id,)
+        )
+        self.render(request)
+        self.assertEqual(channel.code, 200, channel.result)
+        self.assertEqual(channel.json_body["displayname"], "new display name")
```
Comment (anoadragon453) [inline tests/rest/client/test_shadow_banned.py:211]: Can you stick the user in a room and make sure their `m.room.member` event still contains the old displayname?

**Review by anoadragon453 — CHANGES_REQUESTED** [review]

File: tests/rest/client/test_shadow_banned.py (hunk 4 above)
Comment (clokep) [reply tests/rest/client/test_shadow_banned.py:211]: Yes! I wasn't sure how to assert that. Thank you! 👍

Comment (anoadragon453) [reply tests/rest/client/test_shadow_banned.py:211]: Yep, no need for syncing or anything, just use [the GET equivalent](https://matrix.org/docs/spec/client_server/unstable#get-matrix-client-r0-rooms-roomid-state-eventtype-statekey).

**Comment by clokep** [comment]
Thanks @anoadragon453! I added handling of `update_membership` to strip out changes to display name / avatar URL and included more tests.

File: tests/rest/client/test_shadow_banned.py [hunk 5]
```diff
@@ -208,4 +217,56 @@ def test_displayname(self):
...
+            "/_matrix/client/r0/rooms/%s/state/m.room.member/%s"
+            % (room_id, self.banned_user_id),
+            {"membership": "join", "displayname": new_display_name},
+            access_token=self.banned_access_token,
+        )
+        self.render(request)
+        self.assertEquals(200, channel.code, channel.result)
+        self.assertIn("event_id", channel.json_body)
+
+        # The display name in the room should not be.
```
Comment (anoadragon453) [inline tests/rest/client/test_shadow_banned.py:259]:
```suggestion
        # The display name in the room should not be changed.
```

**Review by anoadragon453 — APPROVED** [review]
Ahh, doing it off of `allow_per_room_profiles`, of course!

**Pull request merged**